### PR TITLE
Optimize Continue As New And Solve Missed Raised Events

### PR DESCRIPTION
**Pull request by adarsh1**
This PR aims to optimize ContinueAsNew so that an orchestration continues in memory in the dispatcher if it can instead of enqueueing a new task message.

With this Events Raised After a continueAsNew has been executed will also be collected and carried over to the next iteration of the orchestration. This a feature that can be opted out of by specifying SkipEventsOnContinuation as true in the OrchestrationService.

### CODE REVIEW

**Comment by adarsh1** [comment]
@cgillum @simonporter

File: Test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs [hunk 1]
```diff
@@ -609,7 +605,7 @@ public async Task ActorOrchestrationDeleteAllLargeMessageBlobs(bool enableExtend
 
             // Ideally there would only be three blobs at the end of the test.
```
Comment (cgillum) [inline Test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs]: Since your fix addresses this comment, can the comment be removed?

File: Test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs [hunk 2]
```diff
@@ -1963,7 +1959,7 @@ async Task<string> WaitForOperation()
                 public override void OnEvent(OrchestrationContext context, string name, string input)
                 {
                     Assert.AreEqual("operation", name, true, "Unknown signal recieved...");
-                    if (this.waitForOperationHandle != null)
+                    if (this.waitForOperationHandle != null && !this.waitForOperationHandle.Task.IsCompleted)
```
Comment (cgillum) [inline Test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs]: Is this change actually necessary? If so, why?

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs [hunk 3]
```diff
@@ -816,7 +824,7 @@ async Task AbandonAndReleaseSessionAsync(OrchestrationSession session)
             }
 
             session.StartNewLogicalTraceScope();
-            OrchestrationRuntimeState runtimeState = workItem.OrchestrationRuntimeState;
+            OrchestrationRuntimeState runtimeState = newOrchestrationRuntimeState??workItem.OrchestrationRuntimeState;
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs]: Please add spaces around the `??` operator.

File: src/DurableTask.Core/Command/OrchestrationCompleteOrchestratorAction.cs [hunk 4]
```diff
@@ -11,10 +11,17 @@
 //  limitations under the License.
 //  ----------------------------------------------------------------------------------
 
+using System.Collections.Generic;
+using DurableTask.Core.History;
```
Comment (cgillum) [inline src/DurableTask.Core/Command/OrchestrationCompleteOrchestratorAction.cs]: The convention for code in the DTFx repo is to put `using` statements inside the `namespace` declaration.

File: src/DurableTask.Core/TaskOrchestrationContext.cs [hunk 5]
```diff
@@ -35,6 +35,12 @@ internal class TaskOrchestrationContext : OrchestrationContext
         bool executionTerminated;
         int idCounter;
 
+        public bool HasContinueAsNew => continueAsNew != null;
+
+        public void AddEventToNextIteration(HistoryEvent he){
```
Comment (cgillum) [inline src/DurableTask.Core/TaskOrchestrationContext.cs]: Add a newline before the `{`.

File: src/DurableTask.Core/Command/OrchestrationCompleteOrchestratorAction.cs [hunk 6]
```diff
@@ -24,5 +31,7 @@ internal class OrchestrationCompleteOrchestratorAction : OrchestratorAction
         public string Details { get; set; }
 
         public string NewVersion { get; set; }
+
+        public IList<HistoryEvent> CarryOverEvents { get; set; }
```
Comment (cgillum) [inline src/DurableTask.Core/Command/OrchestrationCompleteOrchestratorAction.cs]: Can this setter be removed?

File: src/DurableTask.Core/TaskOrchestrationExecutor.cs [hunk 7]
```diff
@@ -151,8 +153,15 @@ void ProcessEvent(HistoryEvent historyEvent)
                     this.context.HandleTimerFiredEvent((TimerFiredEvent)historyEvent);
                     break;
                 case EventType.EventRaised:
-                    var eventRaisedEvent = (EventRaisedEvent)historyEvent;
-                    this.taskOrchestration.RaiseEvent(this.context, eventRaisedEvent.Name, eventRaisedEvent.Input);
+                    if (this.skipCarryOverEvents||!this.context.HasContinueAsNew)
```
Comment (cgillum) [inline src/DurableTask.Core/TaskOrchestrationExecutor.cs]: Please add spaces around the `||` operator.

File: test/DurableTask.ServiceBus.Tests/ServiceBusOrchestrationServiceTests.cs [hunk 8]
```diff
@@ -70,6 +71,55 @@ await this.taskHub.AddTaskOrchestrations(typeof(SimplestGreetingsOrchestration))
...
+            // TODO: This requirement may not be ideal and should be revisited.
+            await TestHelpers.WaitForInstanceAsync(this.client, id, 10, waitForCompletion: false);
+
+            OrchestrationInstance temp = new OrchestrationInstance() { InstanceId = id.InstanceId };
+            // Perform some operations
+            await this.client.RaiseEventAsync(temp, "operation", "incr");
+
+            // TODO: Sleeping to avoid a race condition where multiple ContinueAsNew messages
+            //       are processed by the same instance at the same time, resulting in a corrupt
+            //       storage failure in DTFx.
```
Comment (cgillum) [inline test/DurableTask.ServiceBus.Tests/ServiceBusOrchestrationServiceTests.cs]: I suggest removing this TODO

File: Test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs [hunk 9]
```diff
@@ -542,13 +542,9 @@ public async Task ActorOrchestration(bool enableExtendedSessions)
                 // TODO: Sleeping to avoid a race condition where multiple ContinueAsNew messages
                 //       are processed by the same instance at the same time, resulting in a corrupt
                 //       storage failure in DTFx.
```
Comment (cgillum) [inline Test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs]: Can you please remove this TODO?

File: src/DurableTask.ServiceBus/ServiceBusOrchestrationService.cs [hunk 10]
```diff
@@ -753,6 +757,24 @@ public async Task RenewTaskOrchestrationWorkItemLockAsync(TaskOrchestrationWorkI
                             this.ServiceStats.TrackingDispatcherStats.MessageBatchesSent.Increment();
                             this.ServiceStats.TrackingDispatcherStats.MessagesSent.Increment(trackingMessages.Count);
                         }
+
+                        if (newOrchestrationRuntimeState != null && runtimeState != newOrchestrationRuntimeState)
+                        {
+                            trackingMessages = await CreateTrackingMessagesAsync(newOrchestrationRuntimeState, sessionState.SequenceNumber);
+                            TraceHelper.TraceInstance(
+                            TraceEventType.Information,
```
Comment (cgillum) [inline src/DurableTask.ServiceBus/ServiceBusOrchestrationService.cs]: nit: please indent parameters.

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs [hunk 11]
```diff
@@ -845,6 +845,12 @@ async Task AbandonAndReleaseSessionAsync(OrchestrationSession session)
             try
             {
                 session.ETag = await this.trackingStore.UpdateStateAsync(runtimeState, instanceId, executionId, session.ETag);
+
+                if(this.trackingStore is InstanceStoreBackedTrackingStore && workItem.OrchestrationRuntimeState != runtimeState){
+
+                    session.ETag = await this.trackingStore.UpdateStateAsync(workItem.OrchestrationRuntimeState, instanceId, workItem.OrchestrationRuntimeState.OrchestrationInstance.ExecutionId, session.ETag);
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs]: The way this is written, it looks like we might potentially update the tracking store twice for the same instance. I assume this is not the intent, in which case I suggest rewriting this to compute the `runtimeState` and `executionId` parameters outside the `try` block using the conditional logic you've added.

Comment (cgillum) [reply src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs]: Also, please use proper formatting / whitespace.

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs [hunk 12]
```diff
@@ -837,6 +845,12 @@ async Task AbandonAndReleaseSessionAsync(OrchestrationSession session)
             try
             {
                 session.ETag = await this.trackingStore.UpdateStateAsync(runtimeState, instanceId, executionId, session.ETag);
+
+                if(this.trackingStore is InstanceStoreBackedTrackingStore && workItem.OrchestrationRuntimeState != runtimeState){
```
Comment (timjl) [inline src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs]: nit: minor formatting updates
```suggestion
                if (this.trackingStore is InstanceStoreBackedTrackingStore && workItem.OrchestrationRuntimeState != runtimeState)
                {
```

File: src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs (hunk 11 above)
Comment (adarsh1) [reply src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs]: This is the intent as incase of instance store based trackingstores that do not overwrite their state we do need to update that the old execution was continued as new

File: src/DurableTask.Core/Command/OrchestrationCompleteOrchestratorAction.cs [hunk 13]
```diff
@@ -13,8 +13,16 @@
 
 namespace DurableTask.Core.Command
 {
+
```
Comment (timjl) [inline src/DurableTask.Core/Command/OrchestrationCompleteOrchestratorAction.cs]: nit: extra newline

File: src/DurableTask.AzureStorage/Tracking/InstanceStoreBackedTrackingStore.cs [hunk 14]
```diff
@@ -128,7 +128,18 @@ public override Task StartAsync()
         }
 
         /// <inheritdoc />
-        public override async Task<string> UpdateStateAsync(OrchestrationRuntimeState runtimeState, string instanceId, string executionId, string eTag)
+        public override async Task<string> UpdateStateAsync(OrchestrationRuntimeState runtimeState, OrchestrationRuntimeState oldRuntimeState, string instanceId, string executionId, string eTag)
+        {
+            if (runtimeState != oldRuntimeState)
+            {
+                eTag = await UpdateStateAsync(oldRuntimeState, instanceId, oldRuntimeState.OrchestrationInstance.ExecutionId, eTag);
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/Tracking/InstanceStoreBackedTrackingStore.cs]: Please add some comments explaining the double-commit.

File: src/DurableTask.Core/IOrchestrationService.cs [hunk 15]
```diff
@@ -97,6 +97,11 @@ public interface IOrchestrationService
         /// </summary>
         int MaxConcurrentTaskOrchestrationWorkItems { get; }
 
+        /// <summary>
+        ///  Will not carry over unexecuted raised events to the next iteration of an orchestration on ContinueAsNew
+        /// </summary>
+        bool SkipEventsOnContinuation { get; }
```
[the remark this answers is not in the document]
Comment (adarsh1) [reply src/DurableTask.Core/IOrchestrationService.cs]: Sure, Current behavior is Ignore though, do we really want an Error?

File: src/DurableTask.AzureStorage/Tracking/ITrackingStore.cs [hunk 16]
```diff
@@ -65,10 +65,11 @@ interface ITrackingStore
         /// Update State in the Tracking store for a particular orchestration instance and execution base on the new runtime state
         /// </summary>
         /// <param name="runtimeState">The New RuntimeState</param>
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/Tracking/ITrackingStore.cs]: Can we rename `runtimeState` to `newRuntimeState` now that we've added an `oldRuntimeState`?
